lnrpc/walletrpc/walletkit_server.go

```diff
@@ -1625,6 +1636,220 @@ func parseAddrType(addrType AddressType,
+
+	err = wire.WriteVarString(&buf, 0, string(req.Msg))
+	if err != nil {
+		return nil, err
+	}
+
+	digest := chainhash.DoubleHashB(buf.Bytes())
+
+	// For all address types (P2WKH, NP2WKH,P2TR) the edsca compact singing
+	// algorithm is used. For P2TR addresses this represents a special case.
+	// Edsca is used to create a compact signature which makes the public
+	// key of the signature recoverable. For Schnorr no known compact
+	// signing algorithm exists yet. This could change in the future, one
+	// could use the external public key but it was decided to use the
+	// internal private key to sign the message.
+	privKey, err := pubKey.PrivKey()
+	if err != nil {
+		return nil, fmt.Errorf("no private key could be "+
+			"fetched from wallet database %v", err)
+	}
+
+	sigBytes, err := ecdsa.SignCompact(privKey, digest, pubKey.Compressed())
+	if err != nil {
+		return nil, fmt.Errorf("failed to create signature"+
+			"%v", err)
+	}
+
+	// Bitcoin signatures are base64 encoded (being compatible with
+	// bitcoin-core and btcd)
+	sig := base64.StdEncoding.EncodeToString(sigBytes)
+
+	return &SignMessageWithAddrResponse{
+		Signature: sig}, nil
+}
+
+// VerifyMessageWithAddr verifies a signature on a message with a provided
+// address, it checks both the validity of the signature itself and then
+// verifies whether the signature corresponds to the public key of the
+// provided address. There is no dependence on the private key of the address
+// therefore also external addresses are allowed to verify signatures.
+// Supported address types are P2PKH, P2WPKH, NP2WPKH, P2TR.
+func (w *WalletKit) VerifyMessageWithAddr(ctx context.Context,
+	req *VerifyMessageWithAddrRequest) (*VerifyMessageWithAddrResponse,
+	error) {
+
+	sig, err := base64.StdEncoding.DecodeString(req.Signature)
+	if err != nil {
+		return nil, fmt.Errorf("malformed base64 encoding of "+
+			"the signature %v", err)
+	}
+
+	// To prevent attack vectors e.g. signing an already signed signature
+	// a prefix is attached (as bitcoin-core and btcd does).
+	var buf bytes.Buffer
+	err = wire.WriteVarString(&buf, 0, messageSignatureHeader)
+	if err != nil {
+		return nil, err
+	}
+
+	err = wire.WriteVarString(&buf, 0, string(req.Msg))
+	if err != nil {
+		return nil, err
+	}
+
+	expectedMessageHash := chainhash.DoubleHashB(buf.Bytes())
+
+	pk, wasCompressed, err := ecdsa.RecoverCompact(sig,
+		expectedMessageHash)
+	if err != nil {
+		return nil, fmt.Errorf("unable to recover public key "+
+			"from compact signature %v", err)
+	}
+
+	var serializedPubkey []byte
+	if wasCompressed {
+		serializedPubkey = pk.SerializeCompressed()
+	} else {
+		serializedPubkey = pk.SerializeUncompressed()
+	}
+
+	addr, err := btcutil.DecodeAddress(req.Addr, w.cfg.ChainParams)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode address %w", err)
+	}
+
+	if !addr.IsForNet(w.cfg.ChainParams) {
+		return nil, fmt.Errorf("encoded address is for"+
+			"the wrong network %v", addr.EncodeAddress())
```

Comment:
nit: pass `req.Addr` here. I think that goes for all `addr.EncodeAddress()` below.